libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1027 @@
+}
+
+static gboolean
+architecture_skip_filter_applies (JsonNode  *filter_value,
+                                  gboolean  *is_filtered,
+                                  GError   **error)
+{
+  JsonArray *skip_architectures_array = json_node_get_array (filter_value);
+  g_autoptr(GList) architecture_nodes = NULL;
+  g_autofree gchar *node_str = json_node_to_string (filter_value);
+  GList *iter = NULL;
+
+  g_return_val_if_fail (is_filtered != NULL, FALSE);
+
+  if (skip_architectures_array == NULL)
+    {
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected '~architectures' filter to be an array, was: %s",
+                   node_str);
+      return FALSE;
+    }
+
+  architecture_nodes = json_array_get_elements (skip_architectures_array);
+
+  for (iter = architecture_nodes; iter != NULL; iter = iter->next)
+    {
+      const gchar *string = json_node_get_string (iter->data);
+
+      if (string == NULL)
+        {
+          g_set_error (error,
+                       EOS_UPDATER_ERROR,
+                       EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                       "Expected '~architectures' array to contain string values, was: %s",
+                       node_str);
+          return FALSE;
+        }
+
+      if (g_strcmp0 (string, eos_updater_get_system_architecture_string ()) == 0)
+        {
+          *is_filtered = TRUE;
+          return TRUE;
+        }
+    }
+
+  return TRUE;
+}
+
+static gboolean
+action_filter_applies (JsonObject   *object,
+                       const gchar  *filter_key_name,
+                       gboolean     *is_filtered,
+                       GError      **error)
+{
+  g_autoptr (JsonNode) filter_value = json_object_get_member (object, filter_key_name);
+
+  g_return_val_if_fail (is_filtered != NULL, FALSE);
+
+  if (g_strcmp0 (filter_key_name, "~architectures") == 0)
+    return architecture_skip_filter_applies (filter_value, is_filtered, error);
+
+  g_set_error (error,
+               EOS_UPDATER_ERROR,
+               EOS_UPDATER_ERROR_UNKNOWN_ENTRY_IN_AUTOINSTALL_SPEC,
+               "Unknown action filter value '%s', expected one of '~architectures'",
+               filter_key_name);
+  return FALSE;
+}
+
+/* We do this at the same time as reading the JSON node so that we don't have
+ * to keep filter information around in memory */
+static gboolean
+action_node_should_be_filtered_out (JsonNode  *node,
+                                    gboolean  *is_filtered,
+                                    GError   **error)
+{
+  JsonObject *object = json_node_get_object (node);
+  g_autoptr(JsonNode) filters_object_node = json_object_get_member (object, "filters");
+  JsonObject *filters_object = NULL;
+  g_autoptr(GList) filters_object_keys = NULL;
+  GList *iter = NULL;
+
+  g_return_val_if_fail (is_filtered != NULL, FALSE);
+
+  *is_filtered = FALSE;
```

Comment:
The convention is that out values only are set when returning `TRUE` for failable functions.